### PR TITLE
adds support for restful JSON responses

### PR DESCRIPTION
**Pull request by flyingzumwalt**

### CODE REVIEW

File: app/controllers/concerns/curation_concerns/application_controller_behavior.rb [hunk 1]
```diff
@@ -7,20 +7,65 @@ module ApplicationControllerBehavior
...
         session['user_return_to'.freeze] = request.url
-        redirect_to main_app.new_user_session_path, alert: exception.message
+        respond_to do |wants|
+          wants.html { redirect_to main_app.new_user_session_path, alert: exception.message }
+          wants.json { render_json_response(response_type: :unauthorized, message: exception.message) }
+        end
       end
+      #
+      #
+      # if [:show, :edit, :update, :destroy].include? exception.action
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/application_controller_behavior.rb]: Cruft?

File: .rubocop.yml [hunk 2]
```diff
@@ -51,6 +51,7 @@ Metrics/ClassLength:
 
 Metrics/ModuleLength:
   Exclude:
+    - 'app/controllers/concerns/curation_concerns/curation_concern_controller.rb'
```
Comment (cam156) [inline .rubocop.yml]: Curious why this was needed?

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb [hunk 3]
```diff
@@ -46,10 +45,22 @@ def create
...
+        curation_concern = document_list.first
+        raise CanCan::AccessDenied.new(nil, :show) unless curation_concern
+        @presenter = show_presenter.new(curation_concern, current_ability)
+      end
+      wants.json do
+        # load and authorize @curation_concern manually
+        # This has to use #find instead of #load_instance_from_solr because
+        # we want to return values like generic_file_ids in the json
+        @curation_concern = curation_concern_type.find(params[:id]) unless curation_concern
+        authorize! :show, @curation_concern
```
Comment (cam156) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: I would expect the authorization to have already occurred.  Are you expecting different authorization levels for the solr document and the fedora document?

File: app/controllers/curation_concerns/classify_concerns_controller.rb [hunk 4]
```diff
@@ -12,13 +11,12 @@ class ClassifyConcernsController < ApplicationController
...
     end
 
     def create
       classify_concern.attributes = params[:classify_concern]
       if classify_concern.valid?
-        respond_with(classify_concern) do |wants|
+        respond_to do |wants|
           wants.html do
             redirect_to new_polymorphic_path(
               [:curation_concerns, classify_concern.curation_concern_class]
```
Comment (cam156) [inline app/controllers/curation_concerns/classify_concerns_controller.rb]: Should there be a json response here?

File: config/locales/curation_concerns.en.yml [hunk 5]
```diff
@@ -39,6 +39,23 @@ en:
           tag: "could not be updated. You do not have sufficient privileges to edit them."
         title:  "Files failed"
         subject: "Batch upload permission denied"
+    api:
```
Comment (cam156) [inline config/locales/curation_concerns.en.yml]: :clap: for I18N instead of hard coded values!

File: .rubocop.yml (hunk 2 above)
Comment (flyingzumwalt) [inline .rubocop.yml]: It started failing that check.  I didn't have time to refactor the module to be shorter or broken up across multiple modules.

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb (hunk 3 above)
Comment (flyingzumwalt) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: I'm double-checking now, but for some unknown reason load_and_authorize_resource isn't doing its job on :show.  @curation_concern is not getting set.

File: app/controllers/curation_concerns/classify_concerns_controller.rb (hunk 4 above)
Comment (flyingzumwalt) [inline app/controllers/curation_concerns/classify_concerns_controller.rb]: I'm not sure that you would ever submit JSON to that endpoint, but I'll add it for consistency.

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb (hunk 3 above)
Comment (flyingzumwalt) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: Yep.  Double-checked.  If you remove those lines, `@curation_concern` never gets loaded and access controls are never enforced.  I don't know why.  The call to `set_curation_concern_type` in the GenericWork class should have set it up but it doesn't seem to work, at least for the :show action.

File: app/controllers/curation_concerns/classify_concerns_controller.rb (hunk 4 above)
Comment (flyingzumwalt) [inline app/controllers/curation_concerns/classify_concerns_controller.rb]: I added that line.  Ready to merge?

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb (hunk 3 above)
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: It's skiped on show, see https://github.com/projecthydra-labs/curation_concerns/blob/json_api/app/controllers/concerns/curation_concerns/curation_concern_controller.rb#L19 Because we want to load it from solr (for html) and not from Fedora.

File: app/controllers/curation_concerns/classify_concerns_controller.rb [hunk 6]
```diff
@@ -12,21 +11,24 @@ class ClassifyConcernsController < ApplicationController
...
             )
           end
+          wants.json { render json: curation_concern, status: :created }
         end
       else
-        respond_with(classify_concern)
+        respond_to do |wants|
+          wants.html { render 'new', status: :unprocessable_entity }
+          wants.json { render_json_response(response_type: :unprocessable_entity, message: curation_concern.errors.messages) }
```
Comment (jcoyne) [inline app/controllers/curation_concerns/classify_concerns_controller.rb]: This line is not covered by a test. Do we need it?

File: app/controllers/curation_concerns/classify_concerns_controller.rb [hunk 7]
```diff
@@ -12,21 +11,24 @@ class ClassifyConcernsController < ApplicationController
...
             redirect_to new_polymorphic_path(
               [:curation_concerns, classify_concern.curation_concern_class]
             )
           end
+          wants.json { render json: curation_concern, status: :created }
         end
       else
-        respond_with(classify_concern)
+        respond_to do |wants|
+          wants.html { render 'new', status: :unprocessable_entity }
```
Comment (jcoyne) [inline app/controllers/curation_concerns/classify_concerns_controller.rb]: This line is not tested do we need it?

File: app/controllers/concerns/curation_concerns/application_controller_behavior.rb [hunk 8]
```diff
@@ -7,20 +7,40 @@ module ApplicationControllerBehavior
...
+          wants.json { render_json_response(response_type: :unauthorized, message: exception.message) }
+        end
       end
     end
+
+    # render a json response for +response_type+
+
+    def render_json_response(response_type: :success, message: nil, options: {})
+      json_body = CurationConcerns::API.generate_response_body(response_type: response_type, message: message, options: options)
+      render json: json_body, status: response_type
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/application_controller_behavior.rb]: Do we want to render the `location:` for creates and updates.  As done here:
https://github.com/rails/jbuilder/blob/master/lib/generators/rails/templates/controller.rb#L37

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb [hunk 9]
```diff
@@ -36,8 +34,9 @@ def create
       after_create_response
     else
       setup_form
-      respond_with(:curation_concerns, curation_concern) do |wants|
+      respond_to do |wants|
         wants.html { render 'new', status: :unprocessable_entity }
+        wants.json { render_json_response(response_type: :unprocessable_entity) }
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: Render the location of the newly created resource in the json.

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb [hunk 10]
```diff
@@ -46,10 +45,23 @@ def create
...
+        curation_concern = document_list.first
+        raise CanCan::AccessDenied.new(nil, :show) unless curation_concern
+        @presenter = show_presenter.new(curation_concern, current_ability)
+      end
+      wants.json do
+        # load and authorize @curation_concern manually because
+        # load_and_authorize_resource is not doing its job. ???
+        # This has to use #find instead of #load_instance_from_solr because
+        # we want to return values like generic_file_ids in the json
+        @curation_concern = curation_concern_type.find(params[:id]) unless curation_concern
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: why an instance variable and not a local?

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb [hunk 11]
```diff
@@ -93,26 +106,15 @@ def setup_form
...
-
     def _prefixes
       @_prefixes ||= super + ['curation_concerns/base']
     end
 
     def after_create_response
-      respond_with(main_app, :curation_concerns, curation_concern)
+      respond_to do |wants|
+        wants.html { redirect_to [main_app, :curation_concerns, curation_concern] }
+        wants.json { render json: curation_concern, status: :created }
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: Rather than rendering the curation_concern itself, shoudln't this be rendering a jbuilder template? I.E.: 
https://github.com/rails/jbuilder/blob/master/lib/generators/rails/templates/show.json.jbuilder

File: app/controllers/concerns/curation_concerns/curation_concern_controller.rb [hunk 12]
```diff
@@ -120,14 +122,18 @@ def after_update_response
       if actor.visibility_changed?
         redirect_to main_app.confirm_curation_concerns_permission_path(curation_concern)
       else
-        respond_with(main_app, :curation_concerns, curation_concern)
+        respond_to do |wants|
+          wants.html { redirect_to [main_app, :curation_concerns, curation_concern] }
+          wants.json { render json: curation_concern }
```
Comment (jcoyne) [inline app/controllers/concerns/curation_concerns/curation_concern_controller.rb]: Render a jbuilder template here please: https://github.com/rails/jbuilder/blob/master/lib/generators/rails/templates/controller.rb#L51

**Comment by jcoyne** [comment]
Please update this PR to use JBuilder. JBuilder acknowledges that there may be more than one JSON representation for an object, and a jbuilder template will represent a single version of that serialization.  This will enable you to hide or add properties to the API response without changing the canonical to_json for the object.

**Comment by jcoyne** [comment]
I noticed that no changes have been made to the GenericFilesController, do those not need to have json too?

**Comment by flyingzumwalt** [comment]
@jcoyne I can't get the jbuilder template to render any content.  Totally perplexed.  For now, I've included a commented-out line where the builder should be called.

I'll be working on json support for generic_files in the morning.

**Comment by jcoyne** [comment]
@flyingzumwalt that is puzzling. Thanks for trying. Now Rubocop is sad though.

**Comment by cam156** [comment]
@flyingzumwalt I'm happy with this once rubocop is...

**Comment by flyingzumwalt** [comment]
I managed to get it to render the jbuilder template.  Caught a bunch of other edge cases and built proper test coverage for GenericFilesController#update.  Now I need to rebase against master and update the uploader.js to handle the new (more verbose) json error messages.

**Comment by flyingzumwalt** [comment]
Ok.  Once Travis is done, this is ready to merge.

**Comment by jcoyne** [comment]
Travis is failing. It appears you should stub out the GenericFileActor in the GenericFilesController test.

**Comment by flyingzumwalt** [comment]
:green_book: :green_apple: :green_heart: :recycle: :+1:

**Comment by jcoyne** [comment]
@flyingzumwalt what about the `location` header for the updates/creates. Then we'd basically have JSON API compliance: http://jsonapi.org/format/

**Comment by jpstroop** [comment]
Just FYI, @flyingzumwalt isn't with us this week :headstone:. Maybe we should merge and file a separate issue re: :arrow_up: ?

**Comment by flyingzumwalt** [comment]
Also it does render a location header in after_update_response and
after_create_response
